### DISCUSSION TITLE
RNFB Phone Auth setup in Expo

### DISCUSSION THREAD

**Discussion by charles-goode** [invertase/react-native-firebase#5787] [Q&A] [answered]
Sorry that this question is partly outside of scope for this project, but I figured only someone familiar with this project would give the most accurate answer.

I'm setting up RNFB in managed expo through a expo-dev-client. More specifically I'm migrating from the JS web SDK for firebase and expo-firebase-recaptcha to using `@react-native-firebase/app`, `@react-native-firebase/auth`, and  `@react-native-firebase/database`. Everything seems to have gone fine with the database, and some of the auth functions (signing in and out with email/password auth works). However, for my app, users onboard starting with a phone authentication flow. When I attempt phone auth I get `Exception: 'Please register custom URL scheme 'app-1-...' in the app's Info.plist file.' was thrown while invoking signInWithPhoneNumber on target RNFBAuthModule with params ...`. I was able to find #557 and #2346 which pointed me towards useful Firebase documentation that I didn't have to see back when I was using expo-firebase-recaptcha. However, I'm caught up on step 2 of configuring recaptcha verification, because it wants me to use Xcode to modify some app config. In expo this type of thing is typically done in app.json, but I don't exactly know what I'm supposed to do in app.json to match the Xcode commands. (Also I don't have a mac, so I can't really experiment with Xcode)

Another point of confusion is that the long URL scheme suggested by the error log doesn't really match the format of the REVERSED_CLIENT_ID which the firebase docs say should be the custom URL scheme.

@barthap I've seen that you're the Expo-RNFB expert. Any ideas?

**Comment by barthap** [comment, marked as answer]
**Update**:
React Native Firebase 16.4.0 now has built-in Expo config plugin for Phone Auth (by #6645) - see [the configuration instruction](https://rnfirebase.io/auth/phone-auth#expo-setup)

--- 
**Old answer**:
I have no experience with phone auth nor expo-firebase-recaptcha, but after taking a quick look at its [source code](https://github.com/expo/expo/tree/master/packages/expo-firebase-recaptcha) I see that it used WebView and Firebase web API.

Regarding the error - it means the URI scheme is not configured. As far as I know, there's a command to add a scheme to an expo project without opening Xcode:
```
npx uri-scheme <your scheme here>
```
It should be run when android/ios directories are already generated, so after running `expo prebuild`. [A documentation for this command is here](https://www.npmjs.com/package/uri-scheme).

I'm afraid I'm not able to help much more atm.

Edit:
> In expo this type of thing is typically done in app.json, but I don't exactly know what I'm supposed to do in app.json to match the Xcode commands.

Everything what is done in `app.json`, is then applied to native Xcode via config plugins. Some plugins may be found in [packages](https://github.com/expo/expo/tree/master/packages)/expo-*/plugin directory, others can be found in expo-cli repo, e.g. in [prebuild-config package](https://github.com/expo/expo-cli/tree/master/packages/prebuild-config/src)

**Reply by charles-goode** [reply]
Thanks for the quick reply!

Since I don't have a mac, I don't think I can `expo prebuild` (no access to xcode / cocoapods for compilation. I rely on building in the cloud with EAS), so `uri-scheme` is likely out of the question for me. 

With your edit, are you suggesting I dig through those plugins and find where I can adjust the uri schemes and then perhaps make a plugin to control it myself? That's a good thought. Will need more time to pursue that.

I found[ info.plist documentation](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-102207) that corresponds to the xcode UI for adding the uri scheme, so I guess I know what I'm looking for. However, I don't know how to make heads-or-tails of the config plugins.

**Reply by barthap** [reply]
> Since I don't have a mac, I don't think I can expo prebuild (no access to xcode / cocoapods for compilation. I rely on building in the cloud with EAS), so uri-scheme is likely out of the question for me.

Not really 😁 
Prebuild only generates the `ios` directory - it doesn't require Xcode. Then, prebuilt project can be sent to EAS, no longer as "managed", but generic/i don't remember the current name, "bare" anyway.

On the other hand, there must be a way to set a scheme when building managed app with EAS. I'm pretty sure that expo-cli must have a built-in config plugin which takes `scheme` from `app.json` and then runs internally `npx uri-scheme` for it. 
EAS build, under the hood, runs `expo prebuild` anyway to apply the native changes so managed projects so it must apply the scheme too.

**Reply by charles-goode** [reply]
I'll give prebuild a try. Thanks!

**Reply by mikehardy** [reply]
"You might like"... https://github.com/kholia/OSX-KVM
I generally have three OSX "machines" up at any given time for mobile dev, and a windows "machine" for various tihngs, Linux on the metal.

**Reply by barthap** [reply]
I've found a config plugin which gets scheme from app.json and updates Info.plist: https://github.com/expo/expo-cli/blob/master/packages/config-plugins/src/ios/Scheme.ts

It looks like you can [create your own plugin](https://docs.expo.dev/guides/config-plugins/) which looks something like this
```js
// not tested this code, written by hand

// copied this function from code above
function appendScheme(scheme, infoPlist) {
  if (!scheme) {
    return infoPlist;
  }

  const existingSchemes = infoPlist.CFBundleURLTypes;

  return {
    ...infoPlist,
    CFBundleURLTypes: [
      ...(existingSchemes ?? []),
      {
        CFBundleURLSchemes: [scheme],
      },
    ],
  };
}

const withCustomScheme = config => {
  // Ensure the objects exist
  if (!config.ios) {
    config.ios = {};
  }
  if (!config.ios.infoPlist) {
    config.ios.infoPlist = {};
  }

  // Append the scheme
  const myCustomScheme = 'scheme from firebase here';
  config.ios.infoPlist = appendScheme(myCustomScheme, config.ios.infoPlist);

  return config;
};

export default withCustomScheme;
```

**Reply by charles-goode** [reply]
Sidenote: I think `expo config --type introspect` can allow me to see the outcomes of all the plugins which could help.

**Reply by charles-goode** [reply]
My last reply was sent without refreshing and see your code. Looks great! I was just taking the plunge into understanding the plugins, and this makes a lot of sense now. Thanks!

**Reply by charles-goode** [reply]
I might need some help getting across the finish line with writing my own config-plugin. I put it in a file at the root of my project called customURLScheme.js and used `module.exports` to be compliant with the Node.js requirement. I added it to my plugins list and the linter seemed to like it. Yet nothing seems to change in the config when I use expo vscode to preview it. Can't really tell if its being run. I'd really prefer to keep the configuration of this fairly minimal since I'm just trying to see if this is at all possible. The expo-module-scripts thing looks like overkill, but is there some essential step I'd miss that preventing things from working?

**Reply by barthap** [reply]
The problem may be vscode expo - it didn't always preview changes properly for me.
What I do:
1. Git commit first
2. Run prebuild with my custom plugin enabled
3. View the git diff

**Reply by charles-goode** [reply]
When I prebuild, it doesn't make an ios folder :/. Only android folder and the index.js and react-native.config.js files. These don't seem to give me an idea as to what the final config is

**Reply by charles-goode** [reply]
Well I went ahead and built fully and it seems to have worked!! Phone auth on iOS worked!

**Reply by charles-goode** [reply]
Thanks a ton @barthap ! Let me know if I can buy you a drink lol

**Reply by barthap** [reply]
Glad it worked! I've never considered any donation system so there's no easy way to do it 😉

**Reply by mikehardy** [reply]
We have a conflg plugin for phone auth for expo as soon as this releases: #6645

**Reply by sirusbaladi** [reply]
> We have a conflg plugin for phone auth for expo as soon as this releases: #6645

are Silent APNs notifications (instead of captcha) included with the plugin for phone auth?
